src/base/Range.hh

```diff
@@ -37,12 +37,138 @@ namespace celeritas
  * \endcode
  */
 
+//---------------------------------------------------------------------------//
+/*!
+ * Proxy container for iterating over a range of integral values.
+ *
+ * Here, T can be any of:
+ * - an integer,
+ * - an enum that has a "size_" member, or
+ * - an OpaqueId.
+ *
+ * It is OK to dereference the end iterator! The result should just be the
+ * off-the-end value for the range, e.g. `FooEnum::size_` or `bar.size()`.
+ */
+template<class T>
+class Range
+{
+    using TraitsT = detail::RangeTypeTraits<T>;
+
+  public:
+    //!@{
+    //! Type aliases
+    using const_iterator = detail::range_iter<T>;
+    using size_type      = typename TraitsT::counter_type;
+    using value_type     = T;
+    //@}
+
+    template<class U>
+    using step_type = typename TraitsT::template common_type<U>;
+
+  public:
+    //// CONSTRUCTORS ////
+
+    //! Empty constructor for empty range
+    CELER_FUNCTION Range() : begin_{}, end_{} {}
+
+    //! Construct from stop
+    CELER_FUNCTION Range(T end) : begin_{}, end_(end) {}
+
+    //! Construct from start/stop
+    CELER_FUNCTION Range(T begin, T end) : begin_(begin), end_(end) {}
+
+    //// CONTAINER-LIKE ACCESS ////
+
+    //!@{
+    //! Iterators
+    CELER_FORCEINLINE_FUNCTION const_iterator begin() const { return begin_; }
```

Comment:
I can't refind the discussion but you had earlier argument why the PieSlice should not conform to the STL contracts (i.e. `start/stop` rather than `begin/end`).  What tilted the balance in the other direction?

Reply:
I can look it up, but basically `start()` is `front()  == *begin()`, and `stop()` is `back() + 1 == *end()`. The start/stop is "slice" nomenclature, whereas begin/end is for iterators. Rephrasing this class as a Range makes sit work better with the OpaqueId aspect, which sort of evolved over time.